### ISSUE TITLE
Training Curves

### ISSUE THREAD

**Issue by aidansmyth95** [huggingface/nanoVLM#195]
Hi there,

Does anyone have training curves that actually match what we see in the README? Prefereably the 222M nanoVLM model but at this stage honestly any model will do.

Growing skeptical that the framework works as it is reported and could do with some reassurance from other users who have had recent success with it.

<img width="313" height="194" alt="Image" src="https://github.com/user-attachments/assets/a8afa100-ffa4-47e1-b493-dd5b7f3e8fe2" />

Thus far my training curves have fallen short of this accuracy, and when I run the generate script I get nonsensical output not related to the cat image.

<img width="765" height="457" alt="Image" src="https://github.com/user-attachments/assets/d424a39c-1ada-406d-b57a-6c52fae6f023" />

Note that I am using the v0.1 version of the repo due to my desire to use the smaller model and attempt to recreate the training curve as a sanity check.

<img width="656" height="170" alt="Image" src="https://github.com/user-attachments/assets/8646fc5e-b1e9-4941-a433-6ea1041a18b3" />

I presume that my config is off and that is why I cannot recreate the curve? Any info on this would be very helpful.

Kind regards,
Aidan

**Comment by yunzhongfuchen** [comment]
I have the same issue. I didn't modify anything and ran the default config on a single A100 using version 0.1 (simply running python train.py). The accuracy curve is shown below:

![Image](https://github.com/user-attachments/assets/0a662ddc-8d6e-4269-9da8-453dc1f8b00a)
The test results are as follows:
Input:
  What is this? 

Outputs:
  >> Generation 1: This picture is of outside. In the center there is a black and brown color cat seems to be
  >> Generation 2: Snapping Mahogany Palm Fir is a medium-sized platform palm tree with a large leather
  >> Generation 3: This image is taken outdoors. At the bottom of the image there is a ground with a few stones
  >> Generation 4: At the bottom of the image we can see the ground with a white color, on the ground we
  >> Generation 5: In this image we can see a cat, black and white color is present on the ground. Bes

**Comment by yunzhongfuchen** [comment]
When increasing max_new_tokens (e.g., setting it to 500), the officially released pre-trained model usually stops automatically after generating a complete and coherent response, resulting in natural and appropriately lengthed output. However, my own trained model continues generating text until it reaches the maximum token limit, failing to terminate properly. This leads to excessive redundancy, repetitions, and even meaningless content (such as names, dates, URLs, journal titles, etc.), as shown below:
Loading weights from: lusxvr/nanoVLM-222M



This is the result from the official model:


Input:
  What is this? 

Outputs:
  >> Generation 1:  This is a cat sitting on the floor. I think this is a cat sat facing towards the left side. This is a seed of the awaited fruit. This looks like an object in the flower pot. At the top right corner of the image there is another cat sitting on the floor. This is the ground. The background looks slightly blurred.
  >> Generation 2:  In the middle of the picture we can see there is a cream color of cat. We can see a white and black color combination. In the background we can see the path. At the top there is a plant which is in yellow color. On the right side of the picture it looks like a tree. The background of the picture is blur.
  >> Generation 3:  In this image I can see the cat which is in brown and white color on the cream color object. I can see the cat is in front of the black color object and the cat is in the middle of the object. At the top of the image I can see the plants and flowers and I can see another cat is right to the cat.
  >> Generation 4:  The picture is taken in a garden. In the background we can see a green garden. In the center of the picture we can see a cat which is in black and brown color. We can see that the cat is sitting on the ground. Towards the left there is a small plant on the ground. The background of the picture is very blurry.
  >> Generation 5:  In the middle of the picture, we see a cat in black and brown color. Beside that, we see flowers and other greenery. In the background, it is white in color and it is blurred and white in color. This picture is a bit dark. There is a disposal bottle in the middle of the picture and there is a text.


This is the result from the my model:


root@dsw-659799-565c88f96b-6txt4:/mnt/workspace/nanoVLM-0.1# python generate.py --checkpoint checkpoints/nanoVLM-222M
Using device: cuda
Loading weights from: checkpoints/nanoVLM-222M

Input:
  What is this? 

Outputs:
  >> Generation 1: This picture is of outside. In the center there is a black and brown color cat seems to be sitting on the ground. At the top there is a man pointing towards the object which seems to be the object seems to be a garbage box, there is a black color object seems to be a car and in the background there is an ever green background.

Tancred Childress  take burgers for Subway
Tancred Childress, a Texas bride groom pianist is signed by Subway ago.[Cavemanhof: Escapist] Paul McCartney and Tancred Childress on AKA – the party ayonteau

The cavemanaboutmecomeme.comitor diy 39-year-old Walker Engled pro-angle, Paul McCartney star alliance coaxdes C land from Down Under (or the Pampardes atabouts). In true early Cial teams style, two complete lives are depicted on each page.

But on the whole, this is an album stars free, and the originals are always priced pretty high. There's a lot of sunlight on the album.Bike in Mine Reader's Book Dual-Party Calendar

Ask any vintage car or vintage bike aficionado and they will tell you the record for most bike/vehicle mix sales is a 60-over 1. But only. . . Not.The Mississippi Theatre Productions - Italian, French and American (IMT) production was recorded internationally. A Miamite piano piece was performed live from 1995 to 1999 by the Chicago Drama Festival.
Cellar sponge, Cheese, Angus, Field Trip and Architecture were among the works uplifted. CATL family history was recorded.Catalog Spotted Nine Montana River Outfitters!

In this picture we can see a lady in the store with ACTION made up of her hair... MOST OFERS. And here is the buffalo buffalo.The Benefits of Bertolt Baker And Direct Album Cover Singer / vinyl: France / Japan Indie Album Art One-Dimensional - 6" - Volume One Volume One
Bertolt Baker CD by Slim Collection.Great Lakes Battlefield Park (Beltane Press.com) (German Edition) by Slim Collection presents a hardcover record of Bertolt Brecht's "The Bra
...

**Comment by yunzhongfuchen** [comment]
I’ve identified the cause of the issue. During data preprocessing, if only a single eos_token is appended to the end of the answer, the model may continue generating meaningless tokens even after outputting an eos_token during inference. To address this, I now append three eos_tokens consecutively:

answer = item['answer'] + self.tokenizer.eos_token * 3

This way, during training, the model learns that a sequence of multiple consecutive eos_tokens indicates a true end of generation. As a result, during inference, after finishing the meaningful content, the model naturally generates a series of eos_tokens and stops properly — avoiding the problem seen with just one eos_token, where the model would otherwise continue producing irrelevant outputs.

Moreover, since the meaningless tokens generated after the eos_token are masked out in the labels (e.g., set to -100), they do not contribute to the loss computation. This ensures stable training and helps improve prediction accuracy.

![Image](https://github.com/user-attachments/assets/8280a097-7426-40fa-9fce-f136e429df4a)

**Comment by LiuCharming** [comment]
Should the modification be made in the dataset code like this?
messages.append({"role": "assistant", "content": text['assistant']+self.tokenizer.eos_token * 3})

**Comment by aidansmyth95** [comment]
@yunzhongfuchen nice find, did you submit a PR with a proposed change that we can test? Or was it just the change you proposed in your previous answer?

```python
answer = item['answer'] + self.tokenizer.eos_token * 3
```

**Comment by aidansmyth95** [comment]
I did some further debugging. While this might help improve the symptom you described above, it actually isn't the solution. The issue is actually in the eos_token_id for the HuggingFaceTB/cosmo2-tokenizer. Let me explain.

So I assume, like me, you were using the config's setting for lm_eos_token_id: int = 0. However, when you look at the tokenizer, you find that the token ID is actually "endoftext" and not "im_end" (token ID 2).

Once I change the experiment to lm_eos_token_id: int = 2, it seems to converge for MMStar much better, peaking at about 35% accuracy. Give it a try and see, I reckon it fixes the issue.
